core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/HDFSFileMoveAction.java

```diff
@@ -0,0 +1,166 @@
+
+/**
+ * Action that moves file(s) within HDFS in the same cluster.
+ * A user must specify file/directory path and destination file/directory path
+ * Optionals include fileRegex
+ */
+@Plugin(type = Action.PLUGIN_TYPE)
+@Name("HDFSFileMoveAction")
+@Description("Action to move files within HDFS")
+public class HDFSFileMoveAction extends Action {
+  private static final Logger LOG = LoggerFactory.getLogger(HDFSFileMoveAction.class);
+
+  private HDFSActionConfig config;
+
+  public HDFSFileMoveAction(HDFSActionConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void run(ActionContext context) throws Exception {
+    Path source = new Path(config.sourcePath);
+    Path dest = new Path(config.destPath);
+    FileSystem fileSystem = source.getFileSystem(new Configuration());
+
+    if (fileSystem.getFileStatus(source).isFile()) { //moving single file
+
+      try {
+        if (!fileSystem.rename(source, dest)) {
+          if (!config.continueOnError) {
+            throw new IOException(String.format("Failed to rename file {} to {}", source.toString(), dest.toString()));
+          }
+          LOG.error("Failed to move file {} to {}", source.toString(), dest.toString());
+        }
+      } catch (IOException e) {
+        if (!config.continueOnError) {
+          throw e;
+        }
+        LOG.error("Failed to move file {} to {}", source.toString(), dest.toString(), e);
+      }
+      return;
+    }
+
+    // Moving contents of directory
+    FileStatus[] listFiles;
+    if (config.fileRegex != null) {
+      PathFilter filter = new PathFilter() {
+        private final Pattern pattern = Pattern.compile(config.fileRegex);
+
+        @Override
+        public boolean accept(Path path) {
+          return pattern.matcher(path.getName()).matches();
+        }
+      };
+
+      listFiles = fileSystem.listStatus(source, filter);
+    } else {
+      listFiles = fileSystem.listStatus(source);
+    }
+
+    if (listFiles.length == 0) {
+      if (config.fileRegex != null) {
+        LOG.warn("Not moving any files of type {} from source {}", config.fileRegex, source.toString());
+      } else {
+        LOG.warn("Not moving any files from source {}", source.toString());
+      }
+    }
+
+    //Moving directory, so destination filePath doesn't have filename
+    dest = new Path(config.destPath);
+    if (!fileSystem.getFileStatus(dest).isDirectory()) { //directory path was not passed
+      throw new IllegalArgumentException(String.format("Destination path %s is not a directory even though source %s " +
+                                                         "is a directory", dest.toString(), source.toString()));
+    }
+
+    for (FileStatus file: listFiles) {
+      source = file.getPath();
+
+      try {
+        if (!fileSystem.rename(source, dest)) {
+          if (!config.continueOnError) {
+            throw new IOException(String.format("Failed to rename file {} to {}", source.toString(), dest.toString()));
```

Comment:
if a file fails to move, we don't try and move files we've already moved back to their original location. Should mention that in the doc.